sympy/polys/matrices/domainmatrix.py

```diff
@@ -30,6 +30,28 @@ def __init__(self, rows, shape, domain):
         self.shape = shape
         self.domain = domain
 
+    def __getitem__(self, key):
+        i, j = key
+        m, n = self.shape
+
+        if not (isinstance(i, slice) or isinstance(j, slice)):
+            if i >= m:
+                raise IndexError("row index out of range")
+            if j >= n:
+                raise IndexError("column index out of range")
+
+            return DomainScalar(self.rep.getitem(i, j), self.domain)
+
+        i, _m = getindex(i, m)
+        j, _n = getindex(j, n)
+        if i >= m:
+            raise IndexError("row index out of range")
+        if j >= n:
+            raise IndexError("column index out of range")
+
+        return self.from_rep(self.rep.extract_slice(slice(i, _m), slice(j, _n)))
```

Comment:
Why can't this directly pass on the slices received in `key`? I don't see what `getindex` is for and I don't think there's any need to raise since slicing does not normally raise.

Reply:
The purpose of `getindex` is to break the slices into start and stop indices s.t. we can pass them as slices to `extract_slice` s.t. they have positive `start` and `stop` values so that `ri, ci` can give us correct ranges.

If we try to omit the `getindex` method, then there are issues s.t. `ri, ci `doesn't give us proper ranges, and in order to handle them, we include several `if` conditions which make the code even more complex.

Raises are to handle what if someone passes a slice which is out of the size of the `DomainMatrix`. like for a `3*3` DomainMatrix, we give `dM[5: , 0]`

Kindly share your opinion.
Thanks.

Reply:
Slicing with out of range indices returns an empty sequence rather than raising:
```python
In [5]: M = Matrix(range(9)).reshape(3, 3)

In [6]: M[5:, 0]
Out[6]: []
```
You can see the same behaviour with list, tuple, numpy array etc.

Reply:
> Slicing with out of range indices returns an empty sequence rather than raising:

Sure, will remove the checks.

Regarding `getindex`, there will be cases failing and in order to fix them, we will have to add conditions to manipulate them, which is what `getindex` does.
```
In [1]: a = range(3)

In [2]: a[:-3]
Out[2]: range(0, 0)

In [3]: a[2]
Out[3]: 2
```
Like in the first example. we see the o/p as the `range(0, 0)`, but what we want is` range(0, 1)`, this can be fixed with an if condition. E.g. `dM[1:, :-3]`
In the second example, we need it to return as `range` rather than an `int`, so we need to convert the `key` to `slice` using if condition before calling `extract_slice`. E.g. `dM[2,:-3]`.
Similarly, there will be other cases as well, thus `getindex` solves it by handling all such cases.

Kindly share your opinion.
Thanks.